plugins/providers/gcs/provider.go

```diff
@@ -0,0 +1,251 @@
+package gcs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"cloud.google.com/go/iam"
+	"github.com/mitchellh/mapstructure"
+	"github.com/odpf/guardian/core/provider"
+	"github.com/odpf/guardian/domain"
+	"github.com/odpf/guardian/utils"
+)
+
+type Provider struct {
+	typeName string
+	Clients  map[string]GcsClient
+	crypto   domain.Crypto
+}
+
+func NewProvider(typeName string, crypto domain.Crypto) *Provider {
+	return &Provider{
+		typeName: typeName,
+		Clients:  map[string]GcsClient{},
+		crypto:   crypto,
+	}
+}
+
+func (p *Provider) CreateConfig(pc *domain.ProviderConfig) error {
+	c := NewConfig(pc, p.crypto)
+
+	if err := c.ParseAndValidate(); err != nil {
+		return err
+	}
+
+	return c.EncryptCredentials()
+}
+
+func (p *Provider) GetResources(pc *domain.ProviderConfig) ([]*domain.Resource, error) {
+	var creds Credentials
+	if err := mapstructure.Decode(pc.Credentials, &creds); err != nil {
+		return nil, err
+	}
+
+	if err := creds.Decrypt(p.crypto); err != nil {
+		return nil, err
+	}
+
+	client, err := p.getGcsClient(creds)
+	if err != nil {
+		return nil, err
+	}
+
+	resourceTypes := []string{}
+	for _, rc := range pc.Resources {
+		resourceTypes = append(resourceTypes, rc.Type)
+	}
+
+	var resources []*domain.Resource
+	ctx := context.Background()
+	projectID := strings.Replace(creds.ResourceName, "projects/", "", 1)
+	buckets, err := client.GetBuckets(ctx, projectID)
+	if err != nil {
+		return nil, err
+	}
+	for _, b := range buckets {
+		bucketResource := b.toDomain()
+		bucketResource.ProviderType = pc.Type
+		bucketResource.ProviderURN = pc.URN
+		if utils.ContainsString(resourceTypes, ResourceTypeBucket) {
+			resources = append(resources, bucketResource)
+		}
+	}
+
+	return resources, nil
+}
+
+func (p *Provider) GetType() string {
+	return p.typeName
+}
+
+func (p *Provider) GrantAccess(pc *domain.ProviderConfig, a *domain.Appeal) error {
+	if err := validateProviderConfigAndAppealParams(pc, a); err != nil {
+		return fmt.Errorf("invalid provider/appeal config: %w", err)
+	}
+
+	permissions, err := getPermissions(pc.Resources, a)
+	if err != nil {
+		return fmt.Errorf("error in getting permissions: %w", err)
+	}
+
+	var creds Credentials
+	if err := mapstructure.Decode(pc.Credentials, &creds); err != nil {
+		return fmt.Errorf("error in decoding credentials%w", err)
+	}
+
+	if err := creds.Decrypt(p.crypto); err != nil {
+		return fmt.Errorf("error in decrypting credentials%w", err)
+	}
+
+	client, err := p.getGcsClient(creds)
+	if err != nil {
+		return fmt.Errorf("error in getting new client: %w", err)
+	}
+
+	user := a.AccountID
+	userType := a.AccountType
+	identity := fmt.Sprintf("%s:%s", userType, user)
+	ctx := context.TODO()
+	if a.Resource.Type == ResourceTypeBucket {
+		b := new(Bucket)
+		if err := b.fromDomain(a.Resource); err != nil {
+			return fmt.Errorf("from Domain func error: %w", err)
+		}
+		for _, p := range permissions {
+			var role iam.RoleName = iam.RoleName(string(p))
```

Comment:
we can just do this right?
```suggestion
			var role = iam.RoleName(string(p))
```
or
```suggestion
			role := iam.RoleName(string(p))
```